### PR TITLE
Allow to add HTTP headers on top of ApolloClient ones

### PR DESCRIPTION
**Pull request by martinbonnin**
See https://github.com/apollographql/apollo-kotlin/issues/4753

### CODE REVIEW

File: libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/ApolloRequest.kt [hunk 1]
```diff
@@ -59,10 +60,19 @@ private constructor(
 
     override var httpHeaders: List<HttpHeader>? = null
 
+    val additionalHttpHeaders: List<HttpHeader>?
```
[the remark this answers is not in the document]
Comment (martinbonnin) [reply libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/ApolloRequest.kt:63]: Yes because `ApolloRequest` is in `apollo-api` and we need to use it [here](https://github.com/apollographql/apollo-kotlin/blob/9f65ed9277702aa84510fca9ed46ab788c4ad1ec/libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/ApolloClient.kt#L139) from `apollo-runtime`

[the remark this answers is not in the document]
Comment (BoD) [reply libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/ApolloRequest.kt:63]: But this one is on the Builder

**Review by BoD — APPROVED** [review]
👍

File: libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/ApolloRequest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinbonnin) [reply libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/ApolloRequest.kt:63]: Ah my bad. Yea maybe not although all the other ones have it. Don't know. Could be useful to get what's inside a builder?

PS: What we definitely don't need though is setters on all other properties.

[the remark this answers is not in the document]
Comment (martinbonnin) [reply libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/ApolloRequest.kt:63]: > Actually I think we should have `private set` on all the `override var`?

Yea exactly. But this is breaking, I didn't want to do it in the same PR in case we want this for 3.7.5. I'll open a follow up PR

**Comment by martinbonnin** [comment]
Bad news everyone, this broke [this test](https://github.com/apollographql/apollo-kotlin/blob/e6c02d6cb9d0a24ef76af237f827d68a4ff57c27/tests/defer/src/jvmTest/kotlin/test/DeferJvmTest.kt#L50) because interceptors work with `ApolloRequest` and they don't know about `headers` vs `additionalHeaders` (which makes sense). I'll revisit.

**Comment by martinbonnin** [comment]
These tests failing lead me to a somewhat better solution leaving the public API untouched 👌 . `ApolloRequest` is the same, it all happens in `ApolloCall` now. Let me know what you think

File: libraries/apollo-tooling/src/main/kotlin/com/apollographql/apollo3/tooling/SchemaDownloader.kt [hunk 2]
```diff
@@ -185,8 +182,23 @@ object SchemaDownloader {
 
   inline fun <reified T> Any?.cast() = this as? T
 
-  private fun getIntrospectionQuery(includeDeprecatedInputFieldsAndArguments: Boolean): String {
-    val includeDeprecated = if (includeDeprecatedInputFieldsAndArguments) "(includeDeprecated: true)" else ""
+  fun getIntrospectionQuery(specVersion: SpecVersion): String {
```
Comment (BoD) [inline libraries/apollo-tooling/src/main/kotlin/com/apollographql/apollo3/tooling/SchemaDownloader.kt:185]: Some unrelated commit got in the way

**Review by BoD — APPROVED** [review]
👍

File: libraries/apollo-tooling/src/main/kotlin/com/apollographql/apollo3/tooling/SchemaDownloader.kt (hunk 2 above)
Comment (martinbonnin) [reply libraries/apollo-tooling/src/main/kotlin/com/apollographql/apollo3/tooling/SchemaDownloader.kt]: Good catch. I just rebased again. Should be good this time.
